### PR TITLE
Fixed #20846 -- Changed User.username max_length to 254

### PR DESCRIPTION
**Pull request by slurms**

### CODE REVIEW

**Comment by slurms** [comment]
I've addressed one of your points in PR #4250 @timgraham, did we want to address the management command too? Should I also squash this into a single commit?

**Comment by timgraham** [comment]
Maybe a class attribute could be added to the management command so it could be easily overridden if a shorter length is desired.
Squashing commits is fine.
Looks like the migration needs to be moved to 0008 so as not to conflict with the one that's been generated in the meantime.

File: docs/releases/1.9.txt [hunk 1]
```diff
@@ -660,6 +660,29 @@ Database backend API
   has been removed. This method served to format dates on Oracle long
   before 1.0, but hasn't been overridden by any core backend in years
   and hasn't been called anywhere in Django's code or tests.
+:attr:`AbstractUser.username <django.contrib.auth.models.User.username>` max_length increased to 254
```
Comment (timgraham) [inline docs/releases/1.9.txt]: add newline above

File: docs/releases/1.9.txt [hunk 2]
```diff
@@ -660,6 +660,29 @@ Database backend API
...
+
+A migration for :attr:`django.contrib.auth.models.User.username` is included.
+If you have a custom user model inheriting from ``AbstractUser``, you will need
+to generate and apply a database migration for your user model.
+
+To preserve the 30 character limit for usernames, use a custom form when
+creating a user or changing usernames::
+
+    class MyUserCreationForm(UserCreationForm):
+        username = forms.CharField(max_length=30, help_text='Required. '
```
Comment (timgraham) [inline docs/releases/1.9.txt]: please use hanging indent style:
```
username = forms.CharField(
    max_length=30,
    help_text='Required. ',
)
```

File: docs/releases/1.9.txt [hunk 3]
```diff
@@ -660,6 +660,29 @@ Database backend API
...
+            '30 characters or fewer. Letters, digits and @/./+/-/_ only.')
+
+If you wish to keep this restriction in the admin, set
+:attr:`UserAdmin.add_form <django.contrib.auth.admin.UserAdmin.add_form>`
+to use this form.
```
Comment (timgraham) [inline docs/releases/1.9.txt]: . -> ::

File: django/contrib/auth/models.py [hunk 4]
```diff
@@ -301,11 +301,9 @@ class AbstractUser(AbstractBaseUser, PermissionsMixin):
 
     Username and password are required. Other fields are optional.
     """
-    username = models.CharField(
-        _('username'),
-        max_length=30,
-        unique=True,
-        help_text=_('Required. 30 characters or fewer. Letters, digits and @/./+/-/_ only.'),
+    username = models.CharField(_('username'), max_length=254, unique=True,
```
Comment (timgraham) [inline django/contrib/auth/models.py]: please keep hanging indent style

File: django/contrib/auth/migrations/0007_alter_user_username_max_length.py [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+
+    dependencies = [
+        ('auth', '0006_require_contenttypes_0002'),
+    ]
+
+    operations = [
+        migrations.AlterField(
+            model_name='user',
+            name='username',
+            field=models.CharField(error_messages={'unique': 'A user with that username already exists.'}, help_text='Required. 254 characters or fewer. Letters, digits and @/./+/-/_ only.', max_length=254, unique=True, validators=[django.core.validators.RegexValidator('^[\\w.@+-]+$', 'Enter a valid username. This value may contain only letters, numbers and @/./+/-/_ characters.', 'invalid')], verbose_name='username'),
```
Comment (timgraham) [inline django/contrib/auth/migrations/0007_alter_user_username_max_length.py]: please limit line length to 254 chars (manually or with autopep8) -- there is another ticket in progress to do this for the rest of the code base and then add automated enforcement.

File: django/contrib/auth/migrations/0007_alter_user_username_max_length.py [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+from __future__ import unicode_literals
+
+import django.core.validators
+from django.db import migrations, models
+
+
+class Migration(migrations.Migration):
+
+    dependencies = [
+        ('auth', '0006_require_contenttypes_0002'),
```
Comment (timgraham) [inline django/contrib/auth/migrations/0007_alter_user_username_max_length.py]: update dependency for existing 0007 migration; rename file to 0008.

**Comment by slurms** [comment]
Thanks for the review, I'll fix the points you raise shortly (probably Thursday) and ping you again.

File: django/contrib/auth/models.py (hunk 4 above)
Comment (berkerpeksag) [inline django/contrib/auth/models.py]: I couldn't find a test for max_length of User.username in tests/auth_tests/. Should we add a simple one?

Comment (timgraham) [inline django/contrib/auth/models.py]: I think we typically don't require tests for declarative configurations like that.

**Comment by auvipy** [comment]
rebase this against master 1.10

**Comment by timgraham** [comment]
Updated in https://github.com/django/django/pull/5497.

**Pull request closed without merging**